mesos-client/src/main/scala/com/mesosphere/mesos/client/Session.scala

```diff
@@ -0,0 +1,171 @@
+package com.mesosphere.mesos.client
+import java.net.URL
+
+import akka.{NotUsed, util}
+import akka.actor.{Actor, ActorRef, ActorSystem, Props, Stash, Status}
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model.headers.{Authorization, HttpCredentials}
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.settings.ConnectionPoolSettings
+import akka.stream.Materializer
+import akka.stream.scaladsl.Flow
+import com.typesafe.scalalogging.StrictLogging
+
+import scala.concurrent.duration._
+import scala.util.{Failure, Success}
+
+/**
+  * A session captures the connection information with Mesos. This is among other things the URL and credentials providers.
+  *
+  * It behaves similar to Python Request's [[https://2.python-requests.org/en/master/user/advanced/#session-objects Session object]].
+  * Thus it provides methods to construct and connect to Mesos.
+  *
+  * @param url The Mesos master URL.
+  * @param streamId The Mesos stream ID. See the [[http://mesos.apache.org/documentation/latest/scheduler-http-api/#calls docs]] for details.
+  * @param authorization A [[CredentialsProvider]] if the connection is secured.
+  */
+case class Session(url: URL, streamId: String, authorization: Option[CredentialsProvider] = None) {
+  lazy val isSecured: Boolean = url.getProtocol == "https"
+  lazy val port = if (url.getPort == -1) url.getDefaultPort else url.getPort
+
+  /**
+    * Construct a new [[HttpRequest]] for a serialized Mesos call and a set of authorization, ie session token.
+    * @param bytes The bytes of the serialized Mesos call.
+    * @param maybeCredentials The session token if required.
+    * @return The [[HttpRequest]] with proper headers and body.
+    */
+  def createPostRequest(bytes: Array[Byte], maybeCredentials: Option[HttpCredentials]): HttpRequest = {
+    println(s"Body: ${bytes.map(_.toChar).mkString}")
+    HttpRequest(
+      HttpMethods.POST,
+      uri = Uri(s"$url/api/v1/scheduler"),
+      entity = HttpEntity(MesosClient.ProtobufMediaType, bytes),
+      headers = MesosClient.MesosStreamIdHeader(streamId) :: maybeCredentials.map(Authorization(_)).toList
+    )
+  }
+
+  /** @return A flow that makes Mesos calls and outputs HTTP responses. */
+  def post(implicit system: ActorSystem, mat: Materializer): Flow[Array[Byte], HttpResponse, NotUsed] =
+    authorization match {
+      case Some(credentialsProvider) =>
+        implicit val askTimeout = util.Timeout(20.seconds)
```

Comment:
Again, lets put the `20.seconds` in the companion object ;)

Reply:
Also, let's have it configurable via an undocumented environment variable?

Reply:
`reference.conf` it is.